### DISCUSSION TITLE
preferGlobal is broken :/

### DISCUSSION THREAD

**Discussion by RilDev** [npm/feedback#112] [CLI]
Good day,

I recently added `"preferGlobal": true` to my `package.json` but it doesn't add the `-g` flag in the suggested installation snippet on NPM.

![npm-change-default-install-recomendation](https://user-images.githubusercontent.com/13135150/101320727-84c01d00-3864-11eb-9525-a3957ad0812c.png)

I found this Stack-Overflow answer talking about this issue 5 years ago: https://stackoverflow.com/a/28206992

Is there any plan to fix this?

Cordially,
RilDev

**Comment by ljharb** [comment]
Generally speaking, that flag, along with installing anything globally, is discouraged by the ecosystem. It’s not really something you can know with authority that your consumers want to install your package globally - they might have it as a dep or a dev dep, and run it vis an npm run-script, or they might be running it via npx.

**Reply by RilDev** [reply]
Thanks for your reply!

I see, that makes sense. A global install is definitely not optimal in many situations.

In my case though, I get an error when trying to use a local install of this package. And I'd like to make sure it is beginner friendly.

I understand that the `preferGlobal` flag has its drawbacks, but since its already available, wouldn't it make more sense to enable it all the way? And display the expected `npm i [package-name] -g` on the NPM website?

**Reply by ljharb** [reply]
Locally installed packages are run via `npx`, or inside an `npm run-script`, not directly.

**Reply by RilDev** [reply]
That explains why I started having trouble all of a sudden!

Could you link a place in the docs where I could learn more about running installed packages?

**Reply by ljharb** [reply]
https://2ality.com/2016/01/locally-installed-npm-executables.html

**Reply by RilDev** [reply]
Thank you for taking the time to clarify this one for me! I did some testing and I get it now.

For reference:

- `npm i -g gh-packages` will give use access to the `gh-packages` cli on the whole system.
- `npx gh-packages [custom options]` to run the package without having to install it.
- `npm i gh-packages` or `npm i -D gh-packages` will enable us to create a script in `package.json` like so:
```
"scripts": {
  "package": "gh-packages [custom options]"
}
```
This script can be used with `npm run package` or `npm run package -- [additional custom options]`.

**Reply by MylesBorins** [reply]
Small life hack.

If you add `export PATH="$PATH:node_modules/.bin"` to your `.bash_profile` (or equivalent) then you can use all the binaries installed by dependencies in the root of a project.

https://github.com/MylesBorins/dot-files/blob/main/.bash_profile#L10-L11

*note:

It is important to append it to your path (add to the end) rather than prepend (add to beginning) to ensure that npm installed dependencies won't shadow local binaries

**Reply by ruyadorno** [reply]
to build on top of @MylesBorins comment, in `npm@7.1.0` we just added a new interactive subshell that will also enable this workflow  (for the duration of that subshell session, so Myles' life hack is still very relevant if you want to _always_ have binaries available) without the need of tweaking your `.bashrc` file 😊

Running `npm exec` (with no args) will now open a new subshell with `node_modules/.bin` appended to your `$PATH` so that you can run all binaries you regularly can with `npm run` from an interactive shell 😄 give it a try:

```
npm install -g npm@7
```

**Reply by RilDev** [reply]
Thank you @MylesBorins! That little hack makes the `npm` local install work like I expected it to work by default:

```
npm i -D [package-name] # To install package locally as a DevDependency
[package-name] # To run the locally installed package
``` 

@ruyadorno, I tried to use `npm exec` for a while but I couldn't make it work with `nvm`. I tried to reinstall `nvm` and several ways to fix the error like `unset PREFIX` and https://stackoverflow.com/a/47861348/12818567 with closing/reopening my terminal but unsuccessfully :/
Is there anything I missed?
I the meanwhile I opened an issue on `nvm`: https://github.com/nvm-sh/nvm/issues/2379

Have a nice day!
